src/models/tests/restricted_pairwise_union_reach_surface_test.py

```diff
@@ -0,0 +1,129 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tests for restricted_pairwise_union_reach_surface.py."""
+
+from absl.testing import absltest
+import numpy as np
+from wfa_planning_evaluation_framework.models.reach_point import ReachPoint
+from wfa_planning_evaluation_framework.models.reach_surface import ReachSurface
+from wfa_planning_evaluation_framework.models.restricted_pairwise_union_reach_surface import (
+    RestrictedPairwiseUnionReachSurface,
+)
+from wfa_planning_evaluation_framework.models.reach_curve import ReachCurve
+
+
+class LinearCappedReachCurve(ReachCurve):
+    """Linear ReachCurve that is capped at max_reach."""
+
+    def _fit(self) -> None:
+        self._max_reach = self._data[0].reach()
+
+    def by_impressions(self, impressions: [int], max_frequency: int = 1) -> ReachPoint:
+        """Returns the estimated reach for a given impression vector."""
+        kplus_frequencies = [
+            min(sum(impressions), self.max_reach) // i
+            for i in range(1, max_frequency + 1)
+        ]
+        return ReachPoint(impressions, kplus_frequencies)
+
+
+class RestrictedPairwiseUnionReachSurfaceTest(absltest.TestCase):
+    def assertPointsAlmostEqualToPrediction(
+        self, surface, reach_points, tolerance=0.03
+    ):
+        for reach_point in reach_points:
+            prediction = (
+                surface.by_spend(reach_point.spends)
+                if (reach_point.spends)
+                else surface.by_impressions(reach_point.impressions)
+            )
+            self.assertAlmostEqual(
+                prediction.reach(),
+                reach_point.reach(),
+                delta=reach_point.reach() * tolerance,
+            )
+
+    def generate_true_reach(self, a, reach_curves, impressions, spends=None):
+        p = len(reach_curves)
+        reach_vector = [
+            reach_curve.by_impressions([impression]).reach()
+            for reach_curve, impression in zip(reach_curves, impressions)
+        ]
+        reach = sum(reach_vector)
+        for i in range(len(reach_curves)):
+            for j in range(len(reach_curves)):
+                reach -= (a[i * p + j] * reach_vector[i] * reach_vector[j]) / (
+                    max(reach_curves[i].max_reach, reach_curves[j].max_reach) * 2
+                )
+        return ReachPoint(impressions, [reach], spends)
+
+    def generate_sample_reach_curves(self, num_publishers, decay_rate, universe_size):
+        max_reaches = [
+            universe_size * (decay_rate ** pub_num) for pub_num in range(num_publishers)
+        ]
+        return [
+            LinearCappedReachCurve([ReachPoint([max_reach], (max_reach,))])
+            for max_reach in max_reaches
+        ]
+
+    def generate_sample_matrix_a(self, num_publishers):
+        lbd = np.array([[2 / num_publishers] for _ in range(num_publishers)])
```

Comment:
Following what Matthew did before: for testing we may set lbd = 0, i.e., non-overlapping reach surface. The model should have a perfect prediction for this. Would you like to try it?

Reply:
done